lib/nerve.rb

```diff
@@ -15,35 +15,94 @@ class Nerve
 
     include Logging
 
-    def initialize(opts={})
+    def initialize(config_manager)
       log.info 'nerve: starting up!'
+      @config_manager = config_manager
 
       # set global variable for exit signal
       $EXIT = false
 
+      @watchers = {}
+      @watcher_versions = {}
+
+      # Flag to indicate a config reload is required by the main loop
+      # This decoupling is required for gracefully reloading config on SIGHUP
+      # as one should do basically nothing in a signal handler
+      @config_to_load = true
+      # Will be populated by load_config! in the main loop
+      @instance_id = nil
+      @services = {}
+      @heartbeat_path = nil
+
+      Signal.trap("HUP") do
+        @config_to_load = true
+      end
+
+     log.debug 'nerve: completed init'
+    end
+
+    def load_config!
+      log.info 'nerve: loading config'
+      @config_to_load = false
+      @config_manager.reload!
+      config = @config_manager.config
+
       # required options
       log.debug 'nerve: checking for required inputs'
       %w{instance_id services}.each do |required|
-        raise ArgumentError, "you need to specify required argument #{required}" unless opts[required]
+        raise ArgumentError, "you need to specify required argument #{required}" unless config[required]
```

Comment:
Yea, I ended up implementing a config check option similar to apachectl's "configtest"

A restart now looks like:
```
~/pg/nerve(branch:dynamic_reconfiguration*) » bundle exec bin/nerve -c $(CONFIG) -k && kill -SIGHUP $(PID)
I, [2016-03-04T21:29:00.416790 #8134]  INFO -- Nerve::Nerve: nerve: starting up!
I, [2016-03-04T21:29:00.416888 #8134]  INFO -- Nerve::Nerve: nerve: starting run
I, [2016-03-04T21:29:00.416913 #8134]  INFO -- Nerve::Nerve: nerve: loading config
I, [2016-03-04T21:29:00.418033 #8134]  INFO -- Nerve::Nerve: nerve: detected new config for your_http_service
I, [2016-03-04T21:29:00.418087 #8134]  INFO -- Nerve::Nerve: nerve: detected new config for your_tcp_service
I, [2016-03-04T21:29:00.418118 #8134]  INFO -- Nerve::Nerve: nerve: launching new watchers: ["your_http_service", "your_tcp_service"]
I, [2016-03-04T21:29:00.522977 #8134]  INFO -- Nerve::Nerve: nerve: reaping old watchers: []
I, [2016-03-04T21:29:00.523039 #8134]  INFO -- Nerve::Nerve: nerve: configuration check succeeded, exiting
W, [2016-03-04T21:29:00.523064 #8134]  WARN -- Nerve::Nerve: nerve: reaping all watchers
I, [2016-03-04T21:29:00.523083 #8134]  INFO -- Nerve::Nerve: nerve: exiting
exiting nerve
```

This isn't super optimal, but I think it's good enough for this revision.